### PR TITLE
[docs] Document workflow for developing terraform code

### PR DESCRIPTION
**Pull request by Nuru**
## what
Document workflow for developing terraform code
## why
Not previously documented

### CODE REVIEW

**Comment by osterman** [comment]
Please see docs/readme for how to format man page. Mandb expects certain things to be present.

File: docs/developing_terraform_modules.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+available both inside and outside the Geodesic shell.
+ 
+You might create a subdirectory `src` for all your development
+work, and under that have directories for each repo, `terraform-widget`, `terraform-root-modules`
+and one for your Geodesic, something like `prod.cpco.io`.
+
+```
+WORKSTATION:
+
+$HOME=/Users/user
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
$HOME=/Users/john
```

File: docs/developing_terraform_modules.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+/Users/usr/src
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
/Users/john/src
```

File: docs/developing_terraform_modules.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
+## Developing Terraform Modules
+
+This document is not actually about Terraform or how to write good Terraform code or how
```
Comment (osterman) [inline docs/developing_terraform_modules.md]: I would reverse the order of these two statements. First what it is, then what it's not.

File: docs/developing_terraform_modules.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+[build-harness](https://github.com/cloudposse/build-harness) which contains tools for,
+among other things, creating README.md and other documentation files from a combination
+of `yaml` files and the code itself. This process is not well documented and will likely
+undergo revision as we transition to Terraform 0.12, so for now your best bet is to simply
+start with an existing module and make changes to the code to suit your needs. 
+[terraform-aws-key-pair](https://github.com/cloudposse/terraform-aws-key-pair) is a reasonable module to use
+as a starting point, though there are many others that may already be closer to what you want
+to accomplish.
+
+In particular, copy the Makefile and README.yaml files, which will then let you generate
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
In particular, copy the `Makefile` and `README.yaml` files, which will then let you generate 
```

File: docs/developing_terraform_modules.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+The idea of a single module is to do one thing, do it well, and make it easy for users
+to customize by changing inputs, while at the same time making it easy to use by providing
+as many reasonable defaults as possible.
+
+Each Terraform module is placed in its own GitHub repository. CloudPosse has published
+[dozens](https://github.com/cloudposse?utf8=✓&q=terraform&type=&language=) and you can
+follow those examples or use your own pattern. CloudPosse maintains a repo called
+[build-harness](https://github.com/cloudposse/build-harness) which contains tools for,
+among other things, creating README.md and other documentation files from a combination
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
among other things, creating `README.md` and other documentation files from a combination
```

File: docs/developing_terraform_modules.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+infrastructure resources." Each module is intended to provision a set of resources that constitute
+a single logical "thing". For example, a database, which could include AWS instances to run the 
+database software, EBS volumes to store the data, Security Groups and IAM roles and policies to 
+limit network access to the database, etc.
+
+The idea of a single module is to do one thing, do it well, and make it easy for users
+to customize by changing inputs, while at the same time making it easy to use by providing
+as many reasonable defaults as possible.
+
+Each Terraform module is placed in its own GitHub repository. CloudPosse has published
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
Each Terraform module is placed in its own GitHub repository. Cloud Posse has published
```

File: docs/developing_terraform_modules.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+[dozens](https://github.com/cloudposse?utf8=✓&q=terraform&type=&language=) and you can
+follow those examples or use your own pattern. CloudPosse maintains a repo called
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
follow those examples or use your own pattern. Cloud Posse maintains a repo called
```

File: docs/developing_terraform_modules.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+We consider each root module to be a "project" that gets some component installed. It uses
+one or more Terraform modules plus its own additional inputs and logic to create a 
+possibly more complicated component that a single module should handle. Also, while the Terraform modules should do
+one thing, and be fairly generic about how they do it, allowing users to customize it as needed, 
+the Terraform Root Module is the place where opinions are imposed and decisions get made. For 
+example, a module should get all its inputs via variables, but a Root Module may decide to get
+its inputs from Environment variables or AWS SSM Parameter Store parameters or somewhere else.
+
+Each Terraform root module is stored inside its own directory in the shared Terraform
+root module Git repository. CloudPosse's public repo is
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
root module Git repository. Cloud Posse's public repo is 
```

File: docs/developing_terraform_modules.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+documentation by executing:
+```bash
+make init
+make readme/deps
+make readme
+```
+You may need to run that from inside a Geodesic container if you do not want to install
+some of the necessary support (such as the Go language runtime) on your workstation. 
+
+Of course, you will want to modify your README.yaml extensively.
```
Comment (osterman) [inline docs/developing_terraform_modules.md]:
```suggestion
Of course, you will want to modify your `README.yaml` extensively.
```

**Comment by Nuru** [comment]
Superseded by https://github.com/cloudposse/docs/pull/460

**Pull request closed without merging**